### ISSUE TITLE
CPU scaling degrades Actor ping-pong  performance on Linux

### ISSUE THREAD

**Issue by vsarunas** [swiftlang/swift#77947]
### Description

As mentioned by @ktoso [here](https://github.com/swiftlang/swift-corelibs-libdispatch/issues/760#issuecomment-1854815498) and @rjmccall [here](https://github.com/swiftlang/swift-corelibs-libdispatch/issues/760#issuecomment-1855079373); I wanted to demonstrate how not controlling thread count can significantly impact dispatch performance on Linux.

Using a Mac Mini M4 Pro (14-core variant) running Ubuntu 24.04 LTS in a Multipass VM, I tested performance by varying the number of cores available to the VM (as I couldn't find a way to control this via environment variables) and running a basic minimal actor test below.

The results show that while macOS performance remains constant, the same code on Linux degrades as more CPU cores are added. The Linux VM with 14 cores performs several times worse than when run with 2 cores:

<img width="808" alt="linux-core-count-actor" src="https://github.com/user-attachments/assets/7f5cfd66-c932-4faf-8f2d-a58d2b1825f8">

Raw data:
| Pingers | macOS | linux14c | linux12c | linux8c | linux4c | linux2c |
|--------:|------:|----:|----:|---:|---:|---:|
| 1 | 11,846,331 | 11,421,579 | 11,364,196 | 11,869,799 | 11,347,104 | 11,556,880 |
| 2 |3,299,910 | 6,445,558 | 6,123,505 | 3,953,612 | 3,700,211 | 4,939,033 |
| 8 | 2,979,551 | 3,019,615 | 2,465,688 | 2,670,892 | 3,381,066 | 5,511,328 |
| 512 | 3,028,266 | 1,330,248 | 1,366,468 | 1,587,204 | 3,161,267 | 4,942,333 |
| 4,096 | 3,202,390 | 1,264,782 | 1,308,264 | 1,640,846 | 3,294,145 | 5,304,749 |
| 8,192 | 3,017,156 | 1,288,134 | 1,277,824 | 1,581,985 | 3,317,190 | 5,398,518 |
| 16,384 | 2,877,146 | 1,312,917 | 1,307,864 | 1,752,558 | 3,150,326 | 4,416,344 |
| 32,768 | 2,660,722 | 1,273,489 | 1,375,541 | 1,670,825 | 2,765,453 | 3,656,408 |

Performance on x86 processors also degrades as the number of CPU cores in the system increases.

# Stack

[DefaultActorImpl::unlock(bool)](https://github.com/swiftlang/swift/blob/swift-6.0.2-RELEASE/stdlib/public/Concurrency/Actor.cpp#L1811); then into dispatch and futex calls in the kernel:

```
actor-latency-s   33229  3404.987861:     250000 task-clock:ppp: 
        ffff800080132b5c try_to_wake_up+0x28c ([kernel.kallsyms])
        ffff8000801330e4 wake_up_q+0x6c ([kernel.kallsyms])
        ffff8000802064e4 futex_wake+0x1ac ([kernel.kallsyms])
        ffff800080202a24 do_futex+0x144 ([kernel.kallsyms])
        ffff800080202c40 __arm64_sys_futex+0xf8 ([kernel.kallsyms])
        ffff800080032994 invoke_syscall+0x7c ([kernel.kallsyms])
        ffff800080032a94 el0_svc_common.constprop.0+0x4c ([kernel.kallsyms])
        ffff800080032bb0 do_el0_svc+0x28 ([kernel.kallsyms])
        ffff80008165bfc4 el0_svc+0x44 ([kernel.kallsyms])
        ffff80008165c798 el0t_64_sync_handler+0x148 ([kernel.kallsyms])
        ffff800080011648 el0t_64_sync+0x1b0 ([kernel.kallsyms])
            e1c8590bdaf0 sem_post+0x80 (/usr/lib/aarch64-linux-gnu/libc.so.6)
            e1c858f3d550 _dispatch_sema4_signal+0x1c (/usr/local/share/toolchains/6.0.2/usr/lib/swift/linux/libdispatch.so)
            e1c858f36290 _dispatch_semaphore_signal_slow+0x14 (/usr/local/share/toolchains/6.0.2/usr/lib/swift/linux/libdispatch.so)
            e1c858f31f5c _dispatch_root_queue_poke_slow+0x54 (/usr/local/share/toolchains/6.0.2/usr/lib/swift/linux/libdispatch.so)
            e1c859b83a70 (anonymous namespace)::DefaultActorImpl::unlock(bool)+0x164 (/usr/local/share/toolchains/6.0.2/usr/lib/swift/linux/libswift_Concurrency.so)
            e1c859b833e8 swift_task_switchImpl(swift::AsyncContext*, void ( swiftasynccall*)(swift::AsyncContext* swift_async_context), swift::SerialExecutorRef)+0x19c (/usr/local/share/toolchains/6.0.2/usr/lib/swift/linux/libswift_Concurrency.so)
            e1c859b8240c swift::runJobInEstablishedExecutorContext(swift::Job*)+0x19c (/usr/local/share/toolchains/6.0.2/usr/lib/swift/linux/libswift_Concurrency.so)
            e1c859b83728 (anonymous namespace)::ProcessOutOfLineJob::process(swift::Job*)+0x1f0 (/usr/local/share/toolchains/6.0.2/usr/lib/swift/linux/libswift_Concurrency.so)
            e1c859b82398 swift::runJobInEstablishedExecutorContext(swift::Job*)+0x128 (/usr/local/share/toolchains/6.0.2/usr/lib/swift/linux/libswift_Concurrency.so)
            e1c859b82f0c swift_job_run+0x9c (/usr/local/share/toolchains/6.0.2/usr/lib/swift/linux/libswift_Concurrency.so)
            e1c858f2a744 _dispatch_continuation_pop+0xec (/usr/local/share/toolchains/6.0.2/usr/lib/swift/linux/libdispatch.so)
            e1c858f2a574 _dispatch_async_redirect_invoke+0xb8 (/usr/local/share/toolchains/6.0.2/usr/lib/swift/linux/libdispatch.so)
            e1c858f3562c _dispatch_worker_thread+0x1b0 (/usr/local/share/toolchains/6.0.2/usr/lib/swift/linux/libdispatch.so)
```

Strangely `DefaultActorImpl::unlock()` is filtered out here but the rest is intact:

![sample-actor-latency-s-linux-dev-2024-12-04_153310](https://github.com/user-attachments/assets/dd9bae43-871c-42a4-a7b4-5a11a00053db)

### Reproduction

# Steps to reproduce

The example from https://github.com/swiftlang/swift/issues/68299 is suitable for running on Linux with different CPU cores.
<details>
<summary>actor-latency-swift.swift</summary>

```swift

// https://github.com/snaury/coroactors/blob/main/src/comparisons/actor-latency-swift.swift
let clock = ContinuousClock.continuous

actor Pingable {
    private var counter: Int = 0

    func ping() -> Int {
        counter += 1
        return counter
    }   

    func getCounter() -> Int {
        return counter
    }   
}   

actor Pinger {
    private var target: Pingable

    init(_ target: Pingable) {
        self.target = target
    }   

    func run(_ n: Int, start: ContinuousClock.Instant, withLatencies: Bool = true) async -> Duration {
        var maxLatency: Duration = .seconds(0)
        if withLatencies {
            var end = clock.now
            maxLatency = end - start
            for _ in 0..<n {
                let callStart = end
                _ = await self.target.ping()
                let callEnd = clock.now
                let latency = callEnd - callStart
                if maxLatency < latency {
                    maxLatency = latency
                }
                end = callEnd  
            }
        } else {
            for _ in 0..<n {   
                _ = await self.target.ping()
            }
        }
        return maxLatency
    }
}

func main() async {
    var numPingers = 1
    var numPingables = 1
    var count = 10_000_000
    var withLatencies = true   

    var i = 1
    let args = CommandLine.arguments
    while i < args.count {
        switch args[i] {
        case "-p":
            i += 1
            numPingers = Int(args[i])!
            numPingables = numPingers
        case "--pingers":
            i += 1
            numPingers = Int(args[i])!
        case "--pingables":
            i += 1
            numPingables = Int(args[i])!
        case "-c", "--count":  
            i += 1
            count = Int(args[i])!
        case "--without-latencies":
            withLatencies = false
        default:
            precondition(false, "Unexpected argument \(args[i])")
        }
        i += 1
    }

    var pingables = [Pingable]()
    for _ in 0..<numPingables {
        pingables.append(Pingable())
    }
    var pingers = [Pinger]()   
    for i in 0..<numPingers {  
        pingers.append(Pinger(pingables[i % numPingables]))
    }

    //print("Warming up...")   
    let r = count / numPingers / 100
    _ = await pingers[0].run(r, start: clock.now, withLatencies: withLatencies)

    //print("Starting...")
    let start = clock.now
    let maxLatency = await withTaskGroup(of: Duration.self) { group in
        let m = count % numPingers
        for i in 0..<numPingers {
            let n = (count / numPingers) + (i < m ? 1 : 0)
            let pinger = pingers[i]
            let wl = withLatencies
            group.addTask {
                return await pinger.run(n, start: start, withLatencies: wl)
            }
        }

        var maxLatency: Duration = .seconds(0)
        for await latency in group {
            if maxLatency < latency {
                maxLatency = latency
            }
        }
        return maxLatency
    }
    let end = clock.now

    var sum = 0
    for pingable in pingables {
        sum += await pingable.getCounter()
    }
    precondition(sum == r + count)

    let elapsed = end - start  
    let elapsedSeconds = elapsed / Duration.seconds(1)
    let elapsedMs = Int(elapsedSeconds * 1000)
    let rps = Int(Double(count) / elapsedSeconds)
    let maxLatencyUs = maxLatency / Duration.microseconds(1)
    print("Finished in \(elapsedMs)ms (\(rps)/s), max latency = \(maxLatencyUs)us")
}

await main()
```
</details>

<details>
<summary>actor-latency-swift.sh</summary>

```
#!/bin/bash
#export LIBDISPATCH_COOPERATIVE_POOL_STRICT=1

swiftc -swift-version 6 -O actor-latency-swift.swift
for p in 1 2 8 512 4096 8192 16384 32768; do
    echo -n "$p pingers..."
    ./actor-latency-swift --pingables 1 --pingers $p -c 10000000
done
```
</details>

### Expected behavior

Performance should not degrade when running on a host with more CPU cores; it should remain consistent, as it does on macOS

### Environment

```
$ swift --version
Swift version 6.0.2 (swift-6.0.2-RELEASE)
Target: aarch64-unknown-linux-gnu

$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 24.04.1 LTS
Release:        24.04
Codename:       noble
```

**Comment by rjmccall** [comment]
Okay, in the graph we have `n` tasks, each repeatedly jumping back and forth between a task-specific actor and an actor that's shared across all tasks. So it's a baseline expectation that this is going to scale badly because it's heavily contended on the shared actor, and we're trying to decide if this is *surprisingly* bad, even beyond that baseline. This is difficult, because there are a bunch of different reasons this could be scaling badly.

Since this benchmark is fundamentally heavily contended on a single resource, as the number of actual cores trying to do work increases, we do expect the costs of contention to increase, i.e. more time to be wasted trying to perform atomic sequences. So we always need to consider that part of the difference between OSes may just be that macOS Dispatch is more conservative about bringing up new threads and/or that the macOS kernel is more conservative about scheduling those threads onto cores, and that results in less contention because there are fewer cores involved.

We can also see from the stack trace that we're spending a lot of time in Linux's `sem_post`, mostly in the underlying futex implementation in the kernel. When a Swift task leaves an actor that's got more jobs to do, it schedules a job on the global thread pool to keep processing the actor. It's unsurprising that a thread pool would use a condition variable to manage idle threads, but it might be surprising that we're spending quite this much time in the condition variable. This could just be an artifact of the benchmark: we're doing relatively trivial amounts of work, so the heavy contention might just mean that all our jobs are very short outside of the contended sections. I suppose it's also possible that Linux's arm64 futex isn't very well tuned, at least in the kernel used by Ubuntu 24.04. More likely, the thread pool in Dispatch should just be doing a better job of avoiding the condition variable, e.g. by briefly polling the job queue before sleeping.

We could also try to improve throughput at a higher level by having processing jobs stay with the actor (when there are jobs to run there) instead of following the task like they normally do.

**Comment by hassila** [comment]
Thanks for chiming in @rjmccall much appreciated.

We agree that the benchmark is very artificial, it is largely a minimised case highlighting something we see in larger tests, so it is a bit extreme here perhaps, but hopefully still highlights the issue - it is clear the single actor will bound the scalability, we were more concerned about the regression when putting more CPU:s online and the fact that threads seems to be activated/parked too much  during heavier workloads.

> More likely, the thread pool in Dispatch should just be doing a better job of avoiding the condition variable, e.g. by briefly polling the job queue before sleeping.

FWIW, this is exactly what we did a long time ago when helping porting libdispatch to Solaris, we put in a [busy wait on the job queue before sleeping](https://github.com/mheily/libpwq/blob/aad3aa3faeebcd27eb8c29db7b1cf7affd22cfa6/src/posix/manager.c#L431), which made a significant impact in both latency and throughput. We also supported up to N threads spinning as an option, which was useful as we had dedicated machines that only run our software (as is common in many enterprise deployments). Some deployments doesn't care about energy consumption (so much), but would rather trade that for absolute performance.

> We could also try to improve throughput at a higher level by having processing jobs stay with the actor (when there are jobs to run there) instead of following the task like they normally do.

It'd be interesting to see the implications of that for sure.

**Comment by vsarunas** [comment]
Thanks @rjmccall for your detailed reply.

I modified the sample to include busy waiting for specific microsecond intervals. The results revealed the following behaviours. On Linux:
* Dispatch initially creates threads equal to (number of online CPUs + 2)
* For larger workloads (128+ microseconds and over 1 million iterations), dispatch gradually scales down the thread count
* Thread creation is non-deterministic and varies between executions. On a 44-core server, tasks with 8μs of work resulted in anywhere from 12 to 19 threads being created

On macOS dispatch consistently creates 4 threads.

Notably, when the actor's workload is between 8-16 microseconds, the reduced thread count and context switching overhead events out  benchmark performance across all configurations:

| Work Duration (µs) | macOS | Linux 12 cores | Linux 8 cores | Linux 4 cores | Linux 2 cores |
|------------------:|-------|----------------|---------------|---------------|---------------|
| 0 | 2,271,424 (4) | 1,525,708 (14) | 1,666,315 (10) | 2,694,728 (6) | 2,760,723 (4) |
| 1 | 597,680 (4) | 321,175 (14) | 338,059 (10) | 423,585 (6) | 654,123 (4) |
| 2 | 323,101 (4) | 192,648 (14) | 202,441 (10) | 249,950 (6) | 309,921 (4) |
| 4 | 217,865 (4) | 110,849 (14) | 116,749 (10) | 132,877 (6) | 208,692 (4) |
| 8 | 113,926 (4) | 74,747 (13) | 75,848 (10) | 83,902 (6) | 104,977 (4) |
| 16 | 59,694 (4) | 53,217 (8) | 53,052 (8) | 53,300 (6) | 53,947 (4) |
| 32 | 30,542 (4) | 29,256 (6) | 29,285 (7) | 29,305 (6) | 29,303 (4) |
| 64 | 15,390 (4) | 15,094 (6) | 15,104 (6) | 15,088 (6) | 15,094 (4) |

Table contains now: `operations / second (threads count)`

<img width="825" alt="linux-dispatch-with-work-8-pingers" src="https://github.com/user-attachments/assets/de735c17-c32b-40a2-8baf-45322cc45046">

The threads are all created in [_dispatch_root_queue_poke_slow()](https://github.com/swiftlang/swift-corelibs-libdispatch/blob/swift-6.0.2-RELEASE/src/queue.c#L5681):

```
            f04dfc8c5fb8 pthread_create+0x4b8 (/usr/lib/aarch64-linux-gnu/libc.so.6)
            f04dfd8c2048 _dispatch_root_queue_poke_slow+0x140 (/usr/local/share/toolchains/6.0.2/usr/lib/swift/linux/libdispatch.so)
            f04dfd8c4d64 _dispatch_lane_concurrent_drain+0x328 (/usr/local/share/toolchains/6.0.2/usr/lib/swift/linux/libdispatch.so)
            f04dfd8bedfc _dispatch_lane_invoke+0x3f0 (/usr/local/share/toolchains/6.0.2/usr/lib/swift/linux/libdispatch.so)
            f04dfd8c562c _dispatch_worker_thread+0x1b0 (/usr/local/share/toolchains/6.0.2/usr/lib/swift/linux/libdispatch.so)
```


> We could also try to improve throughput at a higher level by having processing jobs stay with the actor (when there are jobs to run there) instead of following the task like they normally do.


This would be great; because currently each actor call is currently a system context switch:
```
$ perf stat -- taskset --cpu-list 0-10,22-32 ./actor-latency-swift --pingables 1 --pingers 8 -w 1 -c 1000000
Finished in 5129ms (194960/s), max latency = 1348.722us

 Performance counter stats for 'taskset --cpu-list 0-10,22-32 ./actor-latency-swift --pingables 1 --pingers 8 -w 1 -c 1000000':

          21520.43 msec task-clock                       #    4.186 CPUs utilized             
            993185      context-switches                 #   46.151 K/sec                     
             66919      cpu-migrations                   #    3.110 K/sec                     
               634      page-faults                      #   29.460 /sec                      
       22405590598      cycles                           #    1.041 GHz                       
       16647573969      instructions                     #    0.74  insn per cycle            
        3044932428      branches                         #  141.490 M/sec                     
          44113655      branch-misses                    #    1.45% of all branches           

       5.141318343 seconds time elapsed

       8.609590000 seconds user
      14.337585000 seconds sys
```

Ubuntu 24.04 uses 6.8 which is a release from March 2024 so relatively fresh.

Updated reproducer:
<details>
<summary>actor-latency-swift.swift</summary>

```swift

// https://github.com/snaury/coroactors/blob/main/src/comparisons/actor-latency-swift.swift

import Foundation

let clock = ContinuousClock.continuous

actor Pingable {
    private var counter: Int = 0
    private var busyWork: UInt64 = 0

    init(busyWork: UInt64) {
        self.busyWork = busyWork
    }

    @inline(__always)
    func busyWait() {
        let target = clock.now + .microseconds(busyWork)
        repeat {
            // busy wait
        } while clock.now < target
    }

    func ping() -> Int {
        counter += 1
        busyWait()
        return counter
    }

    func getCounter() -> Int {
        counter
    }
}

actor Pinger {
    private var target: Pingable

    init(_ target: Pingable) {
        self.target = target
    }

    func run(_ n: Int, start: ContinuousClock.Instant, withLatencies: Bool = true) async -> Duration {
        var maxLatency: Duration = .seconds(0)
        if withLatencies {
            var end = clock.now
            maxLatency = end - start
            for _ in 0 ..< n {
                let callStart = end
                _ = await target.ping()
                let callEnd = clock.now
                let latency = callEnd - callStart
                if maxLatency < latency {
                    maxLatency = latency
                }
                end = callEnd
            }
        } else {
            for _ in 0 ..< n {
                _ = await target.ping()
            }
        }
        return maxLatency
    }
}

func getThreadCount() -> Int {
    #if os(macOS)
        return -1
    #elseif os(Linux)
        guard let stat = try? String(contentsOfFile: "/proc/self/stat", encoding: .utf8),
              let count = Int(stat.split(separator: " ")[19]) else { return -1 }
        return count
    #endif
}

func main() async {
    var numPingers = 1
    var numPingables = 1
    var count = 10_000_000
    var withLatencies = true
    var busyWork: UInt64 = 1

    var i = 1
    let args = CommandLine.arguments
    while i < args.count {
        switch args[i] {
        case "-p":
            i += 1
            numPingers = Int(args[i])!
            numPingables = numPingers
        case "--pingers":
            i += 1
            numPingers = Int(args[i])!
        case "--pingables":
            i += 1
            numPingables = Int(args[i])!
        case "-c", "--count":
            i += 1
            count = Int(args[i])!
        case "-w", "--works-for-micros":
            i += 1
            busyWork = UInt64(args[i])!
        case "--without-latencies":
            withLatencies = false
        default:
            preconditionFailure("Unexpected argument \(args[i])")
        }
        i += 1
    }

    var pingables = [Pingable]()
    for _ in 0 ..< numPingables {
        pingables.append(Pingable(busyWork: busyWork))
    }
    var pingers = [Pinger]()
    for i in 0 ..< numPingers {
        pingers.append(Pinger(pingables[i % numPingables]))
    }

    // print("Warming up...")
    let r = count / numPingers / 100
    _ = await pingers[0].run(r, start: clock.now, withLatencies: withLatencies)

    // print("Starting...")
    let start = clock.now
    let maxLatency = await withTaskGroup(of: Duration.self) { group in
        let m = count % numPingers
        for i in 0 ..< numPingers {
            let n = (count / numPingers) + (i < m ? 1 : 0)
            let pinger = pingers[i]
            let wl = withLatencies
            group.addTask {
                await pinger.run(n, start: start, withLatencies: wl)
            }
        }

        var maxLatency: Duration = .seconds(0)
        for await latency in group {
            if maxLatency < latency {
                maxLatency = latency
            }
        }
        return maxLatency
    }
    let end = clock.now

    var sum = 0
    for pingable in pingables {
        sum += await pingable.getCounter()
    }
    precondition(sum == r + count)

    let elapsed = end - start
    let elapsedSeconds = elapsed / Duration.seconds(1)
    let elapsedMs = Int(elapsedSeconds * 1000)
    let rps = Int(Double(count) / elapsedSeconds)
    let maxLatencyUs = maxLatency / Duration.microseconds(1)
    print("Finished in \(elapsedMs)ms (\(rps)/s), max latency = \(maxLatencyUs)us, actor busy work = \(busyWork)us, threads = \(getThreadCount())")
    // try! await Task.sleep(for: .seconds(60))
}

await main()
```
</details>

<details>
<summary>actor-latency-swift.sh</summary>

```
#!/bin/bash
swiftc -swift-version 6 -O actor-latency-swift.swift
for w in 0 1 2 4 8 16 32 64; do
    echo -n "${w}us work..."
    ./actor-latency-swift --pingables 1 --pingers 8 -w $w -c 1000000
done
```
</details>

**Comment by rjmccall** [comment]
> This would be great; because currently each actor call is currently a system context switch:

Note that we have to add a job to the global thread pool either way, which will perform the same syscall.  Let me break this down.

As a reminder, tasks in your benchmark are pinging back and forth between working on the shared actor and working on an actor that's effectively task-specific.

In the current scheduling, threads in the global pool will typically:
1. get assigned to run a job to process the shared actor
2. dequeue the first task on the shared actor
3. run that task until it requests to switch to the task-specific actor
4. release the shared actor lock, schedule a new job to process the shared actor, and acquire the task-specific actor lock
5. run the task on the task-specific actor until it requests to switch to the shared actor
6. see that there's already a thread running the shared actor, enqueue the task on the actor, release the task-specific actor lock, and then suspend
The job scheduled in step 4 will then do the exact same thing.

If instead the thread followed the actor when there was more work to do, one thread in the global pool would:
1. get assigned to run a job to process the shared actor
2. dequeue the first task on the shared actor
3. run that task until it requests to switch to the task-specific actor
4. see that there are more tasks on the current actor, dequeue the next task, and schedule a new job to process the task-specific actor
5. go to step 3 with the next task

The job scheduled in step would then:
1. get assigned to run a job to process the task-specific actor
2. dequeue the first task on the task-specific actor
3. run that task until it requests to switch to the shared actor
4. see that there's already a thread running the shared actor, enqueue the task on the actor, release the task-specific actor lock, and then suspend

So many of the jobs are actually really short, which is usually bad, but we're at least burning through the actor queue faster because we're not spending any time waiting for a thread to take it over while it has active jobs.
